daemon/src/setup_maker.rs

```diff
@@ -0,0 +1,238 @@
+
+        let (sender, receiver) = mpsc::unbounded();
+        // store the writing end to forward messages from the taker to
+        // the spawned contract setup task
+        self.setup_msg_sender = Some(sender);
+
+        let taker_id = self.taker_id;
+        let contract_future = setup_contract::new(
+            self.taker.sink().with(move |msg| {
+                future::ok(maker_inc_connections::TakerMessage {
+                    taker_id,
+                    msg: wire::MakerToTaker::Protocol { order_id, msg },
+                })
+            }),
+            receiver,
+            (self.oracle_pk, self.announcement.clone()),
+            SetupParams::new(
+                cfd.margin()?,
+                cfd.counterparty_margin()?,
+                cfd.order.price,
+                cfd.quantity_usd,
+                cfd.order.leverage,
+                cfd.refund_timelock_in_blocks(),
+            ),
+            self.build_party_params.clone_channel(),
+            self.sign.clone_channel(),
+            Role::Maker,
+            self.n_payouts,
+        );
+
+        spawn_fallible::<_, anyhow::Error>(async move {
+            let _ = match contract_future.await {
+                Ok(dlc) => this.send(SetupSucceeded { order_id, dlc }).await?,
+                Err(error) => this.send(SetupFailed { order_id, error }).await?,
+            };
+
+            Ok(())
+        });
+
+        Ok(())
+    }
+}
+
+#[xtra_productivity]
+impl Actor {
+    fn handle(&mut self, _msg: Accepted, ctx: &mut xtra::Context<Self>) {
+        let order_id = self.order.id;
+        let this = ctx
+            .address()
+            .expect("actor to be able to give address to itself");
+        let res = self
+            .confirm_order
+            .send(maker_inc_connections::ConfirmOrder {
+                taker_id: self.taker_id,
+                order_id,
+                address: this.clone(),
+            })
+            .await;
+
+        if res.is_err() {
+            let error = anyhow::anyhow!("Failed to deliver order confirmation");
+            tracing::error!(%order_id, "Stopping setup_maker actor: {}", error);
+
+            self.on_completed
+                .send(Completed::Failed { order_id, error })
+                .await
+                .expect("maker_cfd actor to be available")
+                .expect("maker_cfd to handle failed");
+
+            ctx.stop();
+            return;
+        }
+
+        let res = self.contract_setup(this).await;
+
+        if let Err(error) = res {
+            let error = error.context("Failed to start contract setup");
+            tracing::error!(%order_id, "Stopping setup_maker actor: {}", error);
+
+            self.on_completed
+                .send(Completed::Failed { order_id, error })
+                .await
+                .expect("maker_cfd actor to be available")
+                .expect("maker_cfd to handle failed");
+
+            ctx.stop();
+            return;
+        }
```

Comment:
Perhaps worth extracting this into a function that returns `Result` and only doing the `if let Err` dance once.

Reply:
I agree. I wish we had try blocks: https://github.com/rust-lang/rust/issues/31436.

Reply:
In the mean-time, you could abuse an `async` for it :)